### PR TITLE
Load rustdoc's JS search index on-demand.

### PR DESCRIPTION
**Pull request by jsha**
Instead of being loaded on every page, the JS search index is now loaded when either (a) there is a `?search=` param, or (b) the search input is focused.

This saves both CPU and bandwidth. As of Feb 2021, https://doc.rust-lang.org/search-index1.50.0.js is 273,838 bytes gzipped or 2,544,939 bytes uncompressed. Evaluating it takes 445 ms of CPU time in Chrome 88 on a i7-10710U CPU (out of a total ~2,100 ms page reload).

Tested on Firefox and Chrome.

New:
https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html
https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html?search=fn

Old:
https://jacob.hoffman-andrews.com/rust/search-on-load/std/primitive.slice.html
https://jacob.hoffman-andrews.com/rust/search-on-load/std/primitive.slice.html?search=fn

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I have two problems with this change:
 1. The crates list for the search isn't loaded anymore until you focus on the search input.
 2. Your browser is heavily using caching, so once it loaded the search index once, it's never loaded anymore until the hash/URL of the URL changes.

**Comment by jsha** [comment]
> The crates list for the search isn't loaded anymore until you focus on the search input.

Pushed a fix. Thanks for spotting. I've also updated the demo pages at https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html.

> Your browser is heavily using caching, so once it loaded the search index once, it's never loaded anymore until the hash/URL of the URL changes.

In my tests, the browser is spending significant time parsing the JS (and the JSON inside). That's time that blocks other JS work, like building the toggles, which in turn blocks the overall page load from completing. This technique also reduces the memory footprint of doc pages in the common case where someone hasn't used search.

**Comment by GuillaumeGomez** [comment]
When I pick a crate to filter the search results, the picked crate won't show up when I reload the page until I load the search index.

**Comment by GuillaumeGomez** [comment]
Another thing which bothers me is the fact that the crate list arrives quite a while **after** clicking on the menu. It's really bad from a UX perspective...

**Comment by jsha** [comment]
> When I pick a crate to filter the search results, the picked crate won't show up when I reload the page until I load the search index.

Good catch, fixed (and updated the demo pages).

> Another thing which bothers me is the fact that the crate list arrives quite a while after clicking on the menu. It's really bad from a UX perspective...

Agreed, this also bothers me. Perhaps it's possible to generate the crate list statically rather than in JS? I'll take a look.

By the way, this noticeable delay is a good demonstration that loading the search JSON takes non-negligible time even when the JS is cached.

**Comment by GuillaumeGomez** [comment]
> By the way, this noticeable delay is a good demonstration that loading the search JSON takes non-negligible time even when the JS is cached.

That's actually a great argument in favor for this change!

> > Another thing which bothers me is the fact that the crate list arrives quite a while after clicking on the menu. It's really bad from a UX perspective...
> Agreed, this also bothers me. Perhaps it's possible to generate the crate list statically rather than in JS? I'll take a look.

It'd require another JS file to store the crate list but it's possible.

In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).

**Comment by jsha** [comment]
> It'd require another JS file to store the crate list but it's possible.

Is the crate list not known at generation time for any particular page? Also, are there situations where the crate list is quite large, or is the situation on `doc.rust-lang.org` representative (5 crates)?

> In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).

Yep. The issue here is that if you pick a crate, then go back to "All crates", there's still a storage entry. Updated the code to ignore "All crates" (and pushed demo).

**Comment by GuillaumeGomez** [comment]
> Is the crate list not known at generation time for any particular page? Also, are there situations where the crate list is quite large, or is the situation on `doc.rust-lang.org` representative (5 crates)?

It's actually **very** common to have more than 5 crates. To be exact, it depends on how many dependencies you have (so I'll let you do the maths :p).

> Yep. The issue here is that if you pick a crate, then go back to "All crates", there's still a storage entry. Updated the code to ignore "All crates" (and pushed demo).

I precised that my local storage was empty, which meant that I removed the local storage "rustdoc-saved-filter-crate" entry completely through the web dev tools. ;)

**Comment by GuillaumeGomez** [comment]
Sorry, forgot to answer to one of your questions:

> Is the crate list not known at generation time for any particular page? 

No, and the reason is actually a bit more complex than what it seems: you might add more crates to your local documentation but the search index has to work in any case and not simply be overwritten. So to do so, if the search-index already has this crate, it overwrites this entry specifically, otherwise it happens it.

**Comment by jsha** [comment]
I pushed a new version (and updated demo) that generates a separate, smaller, crates.js listing the crates. This is used for building the dropdown and also for adding sidebar items depending on the page. There's still some issue: my crates.js has `alloc`, `core`, `proc_macro`, and `std`. But if I look on doc.rust-lang.org, the list should also include `test`.

> > In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).
> I precised that my local storage was empty, which meant that I removed the local storage "rustdoc-saved-filter-crate" entry completely through the web dev tools. ;)

Found the issue. Looks like a difference between Firefox and Chrome. `savedCrate` was `undefined` in Chrome, but `null` in Firefox.

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -1048,11 +1048,15 @@ themePicker.onblur = handleThemeButtonsBlur;
...
-        // "addSearchOptions" has to be called first so the crate filtering can be set before the
-        // search might start (if it's set into the URL for example).
-        v.push_str("\\\n}');\naddSearchOptions(searchIndex);initSearch(searchIndex);");
+        v.push_str("\\\n}');\ninitSearch(searchIndex);");
         cx.shared.fs.write(&dst, &v)?;
     }
+
+    let crate_list_dst = cx.dst.join(&format!("crates{}.js", cx.shared.resource_suffix));
+    let crate_list =
+        format!("const crates = [{}];", krates.iter().map(|k| format!("\"{}\"", k)).join(","));
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:1057]: `const` isn't part of ES5, please use instead: `window.CRATES = [{}]`. Also: this update is problematic because you remove crates that aren't listed, which is why you don't have `test` listed in the crates. Please update the list instead of rewriting it completely.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1057]: In case you wonder: it's why we update the search index like that and have a crate define on each line separately.

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -2952,6 +2937,35 @@ function defocusSearchBar() {
...
+    window.searchLoaded = false;
+    function loadSearch() {
+        if (!window.searchLoaded) {
+            window.searchLoaded = true;
+            loadScript(window.searchJS);
+        }
+    }
+
+    // Assumes `crates{version}.js` has already been loaded, setting the global const `crates`.
+    addSearchOptions(crates);
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2956]: I don't like much to "assume" things are there in an async world. Instead, please call the function to init the crates lists from the `crates{version}.js` file directly.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2956]: Ok nevermind. It's fine as is. However, please use `window.crates` or equivalent. I don't like the usage of global with such common names, it makes the readibility a lot more problematic.

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -2952,6 +2937,35 @@ function defocusSearchBar() {
...
+    addSidebarCrates(crates);
+    var crateSearchDropDown = document.getElementById("crate-search");
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2958]: Can you put all this in a function too so we can avoid creating variables in the "global" scope of this function please?

**Comment by GuillaumeGomez** [comment]
I think we're getting close to something good! Not having to load the search-index when unneeded is a good idea. I'm still worried about the big delay we have on the first search though... I'll think about that.

**Comment by jsha** [comment]
A couple of thoughts on reducing the feeling of delay on the first search:

Right now, there's no reaction on the rest of the page when you highlight the search box. It would be nice to insert some text below the search box like "Getting ready to search." Alternately: right now the search box says "Click or press 'S' to search, '?' for more options." And it continues to say that even after you focus the search box. We could make it so focusing the search box changes this to say "Type in your search here." That would indicate to the user that their input was received.

Similarly, when the user starts typing, we should be responsive to that typing even if the index is not yet loaded. For instance, We could add text under the search box "Searching for 'yourquery'." One problem: the index JS load is one big event, not interruptible by other events. We could probably work around that by breaking up the JSON.parse into multiple calls (one for each crate), and adding setTimeout between. Similarly, I think there's a fair amount of work happening to turn the parsed JSON into in-memory structures; we could similarly add setTimeout during that work to make it possible to update the page in between.

**Comment by jsha** [comment]
By the way, I was looking at the delay when searching and realized we have an artificial delay of 500ms before any search gets executed: https://github.com/rust-lang/rust/blob/301ad8a4fa3ea56fb980443b7997c8f9d72dd717/src/librustdoc/html/static/main.js#L1935.

**Comment by GuillaumeGomez** [comment]
Yep, but in here we outrun this timeout apparently. ;)

**Comment by jsha** [comment]
I tested the `setTimeout` concept. Unfortunately it didn't seem to help. None of the keyboard input events fired between loading, parsing, and buildIndex.

I did find that `search` was running twice. Once, right after `buildIndex`, it would run with whatever was in the search box, even if I was still typing. Since the search itself is a bit expensive (250ms on my machine), doing it twice slows everything down. I fixed that so that "right after buildIndex" search only fires when there's a search in the URL bar.

Overall, I think this is the best we're likely to do. If we wanted to get really fancy, we could use a Web Worker that always has a copy of the index loaded. That would make searches blazing fast, but it wouldn't work for local files so we'd need two code paths. That's probably too much complexity.

File: src/librustdoc/html/render/mod.rs [hunk 4]
```diff
@@ -1037,22 +1037,27 @@ themePicker.onblur = handleThemeButtonsBlur;
...
-        // "addSearchOptions" has to be called first so the crate filtering can be set before the
-        // search might start (if it's set into the URL for example).
-        v.push_str("\\\n}');\naddSearchOptions(searchIndex);initSearch(searchIndex);");
+        v.push_str("\\\n}');\ninitSearch(searchIndex);");
         cx.shared.fs.write(&dst, &v)?;
     }
+
+    let crate_list_dst = cx.dst.join(&format!("crates{}.js", cx.shared.resource_suffix));
+    let crate_list =
+        format!("window.ALL_CRATES = [{}];", krates.iter().map(|k| format!("\"{}\"", k)).join(","));
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:1058]: I think this is still invalid: you'll overwrite any crate that isn't a dependency of the current one (when you are in a workspace for example, or simply documenting unrelated code in the same location). You need to extract the crate list and then update it like we do in the search index (we don't overwrite, we update, which is very different!).

File: src/librustdoc/html/static/main.js [hunk 5]
```diff
@@ -1917,8 +1918,8 @@ function defocusSearchBar() {
             return searchWords;
         }
 
-        function startSearch() {
-            var callback = function() {
+        function registerSearchEvents() {
+            var searchAfter500ms = function() {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:1922]: That's a nice naming improvement, thanks!

File: src/librustdoc/html/static/main.js [hunk 6]
```diff
@@ -2843,22 +2839,14 @@ function defocusSearchBar() {
...
-        var crates_text = [];
-        if (Object.keys(crates).length > 1) {
-            for (var crate in crates) {
-                if (hasOwnProperty(crates, crate)) {
-                    crates_text.push(crate);
-                }
-            }
-        }
-        crates_text.sort(function(a, b) {
+        crates.sort(function(a, b) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2849]: I just thought: now that the crate list is separated from the search index, it might be nice to sort them when generating the crate list in the rust code instead of doing it client side every time. What do you think? If you think it's too much, please add a FIXME comment so it can be taken care of later on.

File: src/librustdoc/html/static/main.js [hunk 7]
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
...
+    function setupSearchLoader() {
+        var searchLoaded = false;
+        function loadSearch() {
+            if (!searchLoaded) {
+                searchLoaded = true;
+                loadScript(window.searchJS);
+            }
+        }
+
+        // Assumes `crates{version}.js` has already been loaded, setting window.ALL_CRATES.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2956]:
```suggestion
        // `crates{version}.js` should always be loaded before this script, so we can use it safely. Time to set `window.ALL_CRATES`!
```

File: src/librustdoc/html/static/main.js [hunk 8]
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
...
+        addSearchOptions(window.ALL_CRATES);
+        addSidebarCrates(window.ALL_CRATES);
+
+        search_input.addEventListener("focus", function(){
+            search_input.origPlaceholder = search_input.placeholder;
+            search_input.placeholder = "Type your search here.";
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2962]: I'm not sure to understand why you're updating the placeholder when the input gets the focus. Any reason?

File: src/librustdoc/html/static/main.js [hunk 9]
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
...
+                searchLoaded = true;
+                loadScript(window.searchJS);
+            }
+        }
+
+        // Assumes `crates{version}.js` has already been loaded, setting window.ALL_CRATES.
+        addSearchOptions(window.ALL_CRATES);
+        addSidebarCrates(window.ALL_CRATES);
+
+        search_input.addEventListener("focus", function(){
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2960]:
```suggestion
        search_input.addEventListener("focus", function() {
```

File: src/librustdoc/html/static/main.js [hunk 10]
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
...
+            search_input.origPlaceholder = search_input.placeholder;
+            search_input.placeholder = "Type your search here.";
+            loadSearch();
+        });
+        search_input.addEventListener("blur", function(){
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2965]:
```suggestion
        search_input.addEventListener("blur", function() {
```

File: src/librustdoc/html/static/main.js [hunk 11]
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
...
+            search_input.placeholder = search_input.origPlaceholder;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2966]: If replacing the placeholder isn't useful, this callback can be removed.

File: src/librustdoc/html/static/main.js (hunk 6 above)
Comment (jsha) [reply src/librustdoc/html/static/main.js:2849]: Yep, I'd be happy to do this!

File: src/librustdoc/html/static/main.js (hunk 8 above)
Comment (jsha) [reply src/librustdoc/html/static/main.js:2962]: This is part of my attempts to make the search feel more responsive. Even though it might take some time for the search index to load, and for your query to run, it's nice for the page to immediately respond to your input in some way. This is the "response" idea from [Response, Animation, Idle, Load (RAIL)](https://web.dev/rail/).

File: src/librustdoc/html/render/mod.rs (hunk 4 above)
Comment (jsha) [reply src/librustdoc/html/render/mod.rs:1058]: I may be misunderstanding, but I think this does the right thing. the `krates` list comes from `collect_json`, which reads it from the `search-index{}.js` as it currently stands. So as each run of `rustdoc` happens, new crates are added to `search-index{}.js`, and `crates{}.js` is updated based on the latest contents of `search-index{}.js`.

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:1058]: As a test, I checked out https://github.com/servo/rust-url, which is a workspace containing 5 members:

```
[workspace]
members = ["url", "form_urlencoded", "idna", "percent_encoding", "data-url"]
```

Building docs with this branch produces this (truncated) search-index:

```
var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

And this crates.js:
```
window.ALL_CRATES = ["data_url","form_urlencoded","idna","matches","percent_encoding","tinyvec","tinyvec_macros","unicode_bidi","unicode_normalization","url"];
```

Building docs with nightly produces this (truncated) search-index:

```var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1060]: And what happens if you now document another non-related crate and make the output into this folder? The `search-index.js` will be updated and the `crates.js` will very likely only contain this new crate.

Comment (jsha) [reply src/librustdoc/html/render/mod.rs:1060]: This also works. I just tested by doing:

```
cd rust/ureq
cargo doc
cd ../conduit-cookie
cargo doc --target-dir ../ureq/target/
```

The resulting crates.js includes ureq, conduit-cookie, and all the dependencies of each:

```
window.ALL_CRATES = ["aead","aes","aes_gcm","aes_soft","base64","block_buffer","bytes","cfg_if","chunked_transfer","cipher","conduit","conduit_cookie","conduit_middleware","const_fn","cookie","cpuid_bool","crypto_mac","ctr","digest","fnv","form_urlencoded","generic_array","getrandom","ghash","hkdf","hmac","http","idna","itoa","libc","log","matches","once_cell","opaque_debug","percent_encoding","polyval","ppv_lite86","proc_macro2","proc_macro_hack","quote","rand","rand_chacha","rand_core","ring","rustls","sct","sha2","spin","standback","subtle","syn","time","time_macros","time_macros_impl","tinyvec","tinyvec_macros","typenum","unicode_bidi","unicode_normalization","unicode_xid","universal_hash","untrusted","ureq","url","webpki","webpki_roots"];
```

File: src/librustdoc/html/render/mod.rs [hunk 12]
```diff
@@ -1039,22 +1039,29 @@ themePicker.onblur = handleThemeButtonsBlur;
...
     }
 
-    // Update the search index
+    // Update the search index and crate list.
     let dst = cx.dst.join(&format!("search-index{}.js", cx.shared.resource_suffix));
     let (mut all_indexes, mut krates) = try_err!(collect_json(&dst, &krate.name.as_str()), &dst);
     all_indexes.push(search_index);
+    krates.push(krate.name.to_string());
+    krates.sort();
+    krates.dedup();
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:1048]: Just wondering: is that case even possible? Normally you can only have each crate once. Did you found a case where there was a duplicate?

File: src/librustdoc/html/render/mod.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1062]: I just checked, you are absolutely right!

File: src/librustdoc/html/static/main.js (hunk 10 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2972]: Still need to add the missing whitespace. ;)

File: src/librustdoc/html/static/main.js (hunk 8 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2969]: Tested it and I actually like how it reacts so let's keep it! :)

**Comment by GuillaumeGomez** [comment]
I went through the changes once again and I think we're getting close to the approval. Just two comments (one nit and one question about the `dedup`). Thanks for going through this review process! :)

File: src/librustdoc/html/render/mod.rs (hunk 12 above)
Comment (jsha) [reply src/librustdoc/html/render/mod.rs:1048]: I was moving this up from here: https://github.com/rust-lang/rust/pull/82310/files#diff-40a0eb025da61717b3b765ceb7fab21d91af3012360e90b9f46e15a4047946faL1081. I do think it's possible that `krates.push(krate.name.to_string())` adds a crate name that is a duplicate of a crate already in the list.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1048]: Look at the code of `collect_json`, this more precisely:

```rust
// ...
let prefix = format!("\"{}\"", krate);
// ...
    if line.starts_with(&prefix) {
        continue;
    }
// ...
```

If the crate is already in the search index, we don't add it to the crates list returned. So the `dedup` is unnecessary here.

**Comment by jsha** [comment]
Updated, thanks!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+
